rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java

```diff
@@ -514,4 +539,82 @@ else if (Character.isSupplementaryCodePoint(currentChar)) {
             }
         }
     }
+
+    // The idea behind the patch is to dynamically map productions in Rascal's
+    // own grammar with no category (e.g, number literals) to current categories
+    // (i.e., semantic token types).
+    private interface CategoryPatch extends BiFunction<IConstructor, String, String> {};
+
+    private static class DefaultCategoryPatch implements CategoryPatch {
+        @Override
+        public String apply(IConstructor prod, String defaultCategory) {
+            return defaultCategory;
+        }
+    }
+
+    private static class RascalCategoryPatch implements CategoryPatch {
+
+        // As an optimization, productions that have already been mapped in the
+        // past, are cached for fast lookups in the future. There are separate
+        // caches for productions that *do* need to be mapped to a new category
+        // and those that *do not* need this.
+        private Cache<IConstructor, String>  doPatch = Caches.build();
+        private Cache<IConstructor, Boolean> doNotPatch = Caches.build();
+
+        @SuppressWarnings("java:S131") // Switches without defaults are intended in this method
+        @Override
+        public String apply(IConstructor prod, String defaultCategory) {
+
+            // Check the caches
+            var category = doPatch.getIfPresent(prod);
+            if (category != null) {
+                return category;
+            }
+            if (doNotPatch.getIfPresent(prod) != null) {
+                return defaultCategory; // Possibly different each time this cache is hit
+            }
+
+            // Apply the patch
+            var def = ProductionAdapter.getDefined(prod);
+
+            if (isLabeledLiteral(def)) {
+                switch (SymbolAdapter.getLabel(def)) {
+                case "integer":
+                case "real":
+                case "rational":
+                    return Caches.putThenGet(doPatch, prod, SemanticTokenTypes.Number);
+                case "location":
+                    return Caches.putThenGet(doPatch, prod, SemanticTokenTypes.String);
+                case "regExp":
+                    return Caches.putThenGet(doPatch, prod, SemanticTokenTypes.Regexp);
+                }
+            }
+
+            doNotPatch.put(prod, Boolean.TRUE);
+            return defaultCategory;
+        }
+
+        private boolean isLabeledLiteral(IConstructor def) {
+            return SymbolAdapter.isLabel(def) &&
+                SymbolAdapter.getName(SymbolAdapter.getLabeledSymbol(def)).equals("Literal");
+        }
+    }
+
+    private static class Caches {
+
+        public static <K, V> Cache<K, V> build() {
```

Comment:
nit: does the key need to be generic? we're only using it for a very specific case I think?

Reply:
True, it doesn't have to be. But the cost of genericity was really low in this case 😉

I've changed it though, because the comment about the safe usage of weak keys (below) breaks the genericity and is specific to keys being constructors

Reply:
ok, good :+1: